compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -509,7 +509,24 @@ static SmallVector<int64_t> getDefaultDistributedLevelTileSizes(
         getMaxTileSize(lbs[i], ubs[i], distributedTileSizes[i], minTileSizes[i],
                        allowIncompleteTile);
   }
-  return distributedTileSizes;
+  // All higher dimensions should be tiled at the workgroup level
+  // (distributedTileSizes), except after encountering an already specified
+  // workgroup tile - because at that point, parallel and reduction tiling would
+  // handle the required tiling. If the dimensions is less than 3, again
+  // parallel and reduction tiling would handle it.
+  SmallVector<int64_t> updatedDistributedTileSizes;
+  updatedDistributedTileSizes.reserve(distributedTileSizes.size());
+  bool foundNonZero = false;
+  for (auto val : distributedTileSizes) {
+    if (val == 0) {
+      updatedDistributedTileSizes.push_back(
+          (foundNonZero || distributedTileSizes.size() <= 3) ? 0 : 1);
```

Comment:
I dont think 3 is hardcoded anywhere. I use [kMaxNumParallelDims ](https://github.com/iree-org/iree/blob/479c1bba717b1e15f7a4f5d980e7aaf161b069f1/compiler/src/iree/compiler/Codegen/Utils/Utils.h#L26)

Reply:
I understand the `foundNonZero` part, but not the `distributedTileSizes.size <= 3` part....

Reply:
what would happen if we just drop the `distributedTileSizes.size() <= 3` .

Reply:
Other tests fail. But more importantly, things get tiled to 1, when parallel or reduction tiling requires something > 1 to be efficient.

This is definitely a hack, because the assumption is that we have 3 parallel tiling dimensions and 1 reduction tiling dimensions (so it should be 4?)

Reply:
I am saying, just stop on the first non-zero dimension you find? I understand the `foundNonZero` part, not the `distributedTileSizes.size <= 3` part.

Reply:
It should probably be 4. From the comment on line 515:

```
  // ... If the dimensions is less than 3, again
  // parallel and reduction tiling would handle it.
```

The parallel tiling takes care of 3 dimensions starting from the penultimate dimension and the reduction tiling takes care of the last dimension. We don't want to tile those dimensions given that parallel and reduction tiling takes care of them. So it should be 4, not 3.

The alternative is to ensure that the workgroup tiling size is = parallel or reduction tiling size when the others handle it. That is the less hacky way of doing this, but then we have to implement a "correction" after determining the parallel and reduction tile sizes.

All this should go away if the tile size system is overhauled.

Reply:
Do we commit this as is?